src/inmanta/agent/forking_executor.py

```diff
@@ -163,7 +169,40 @@ def connection_lost(self, exc: Exception | None) -> None:
 
 
 class ExecutorClient(FinalizingIPCClient[ExecutorContext], LogReceiver):
-    pass
+    def __init__(self, name: str):
+        super().__init__(name)
+
+        # Keeps track of when this client was active last
+        self.last_used_at: datetime.datetime = datetime.datetime.now().astimezone()
+
+    def get_idle_time(self) -> timedelta:
+        return datetime.datetime.now().astimezone() - self.last_used_at
+
+    @typing.overload
+    def call(
+        self, method: IPCMethod[ExecutorContext, ReturnType], has_reply: typing.Literal[True] = True
+    ) -> Future[ReturnType]: ...
+
+    @typing.overload
+    def call(self, method: IPCMethod[ExecutorContext, ReturnType], has_reply: typing.Literal[False]) -> None: ...
+
+    @typing.overload
+    def call(self, method: IPCMethod[ExecutorContext, ReturnType], has_reply: bool = True) -> Future[ReturnType] | None: ...
+
+    def call(self, method: IPCMethod[ExecutorContext, ReturnType], has_reply: bool = True) -> Future[ReturnType] | None:
+        """Call a method with given arguments"""
+        self.last_used_at = datetime.datetime.now().astimezone()
+        response = super().call(method, has_reply)
+        assert response is None or isinstance(response, Future)
+        return response
+
+    def has_outstanding_calls(self) -> bool:
+        """Is this client still waiting for replies"""
+        return len(self.requests) > 0
+
+    def process_reply(self, frame: IPCReplyFrame) -> None:
+        self.last_used_at = datetime.datetime.now().astimezone()
```

Comment:
I have a slight preference to set this after processing the reply, because that is actually when it was last used imo (the last action was taken then).

Reply:
Maybe we should update it before and after then ? Because if we only do it after we process the reply I think there is a chance that we `del self.requests[frame.id]` before updating `last_used_at` and end up cleaning up the executor wrongfully.

Reply:
I'm fine with before and after, but I don't think there's realistically a problem with doing it just after because we're in a synchronous context here. That's why I asked [this question](https://inmanta.slack.com/archives/CKRF0C8R3/p1719826294775269) on Slack BTW.

Reply:
I see :+1: